src/Bicep.Core/Analyzers/Linter/Rules/WhatIfShortCircuitingRule.cs

```diff
@@ -0,0 +1,194 @@
+                            suppliedParameterValues: moduleParamsInput,
+                            parameterValuesPositionalMetadata: null,
+                            metadata: metadata,
+                            metricsRecorder: null,
+                            skipEvaluationOfNonDeterministicFunctionsForWhatIf: false);
+                    }
+                    catch (Exception ex)
+                    {
+                        // TODO: Raise diagnostic when the template evaluation fails if the diagnostic will provide extra context to the failure
+                        // Adding a diagnostic here without checking for other errors would currently result in duplicate errors for any error in module params
+                        Trace.WriteLine($"Exception occurred while reducing template language expressions: {ex}");
+                        continue;
+                    }
+
+                    var sentinelVisitor = new CrossModuleSentinelVisitor();
+
+                    foreach (var resource in templateResult.Resources)
+                    {
+                        CheckResource(resource, sentinelVisitor);
+                    }
+
+                    var resourceTypeResolver = ResourceTypeResolver.Create(model);
+                    foreach (var param in sentinelVisitor.parametersEncountered)
+                    {
+                        var paramSyntax = moduleParamsHolder[param];
+                        if (!IsDeployTimeConstant(paramSyntax, model, resourceTypeResolver))
+                        {
+                            yield return CreateDiagnosticForSpan(diagnosticLevel, paramSyntax.Value.Span, paramSyntax.Value, module.Name);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static bool IsDeployTimeConstant(ObjectPropertySyntax syntax, SemanticModel model, ResourceTypeResolver resolver)
+        {
+            var diagWriter = ToListDiagnosticWriter.Create();
+            DeployTimeConstantValidator.CheckDeployTimeConstantViolations(syntax, syntax.Value, model, diagWriter, resolver); 
+
+            return diagWriter.GetDiagnostics().Count == 0;
+        }
+
+        private void CheckResource(TemplateResourceWithParsedExpressions resource, CrossModuleSentinelVisitor sentinelVisitor)
+        {
+            if (resource.Resources != null)
+            {
+                foreach (var nestedResource in resource.Resources)
+                {
+                    CheckResource(nestedResource, sentinelVisitor);
+                }
+            }
+            sentinelVisitor.Visit(resource.Name);
+            sentinelVisitor.Visit(resource.SubscriptionId);
+            sentinelVisitor.Visit(resource.ResourceGroup);
+            sentinelVisitor.Visit(resource.Condition);
+            sentinelVisitor.Visit(resource.Scope);
+            sentinelVisitor.Visit(resource.ApiVersion);
```

Comment:
[nit]: Normally you want to reverse this calling pattern for visitors, e.g. `resource.Name?.Accept(sentinelVisitor);` instead of `sentinelVisitor.Visit(resource.Name)`. This would also let you remove the null check in `sentinelVisitor.Visit` and make it a private method.